npm/pkg/dataplane/ipsets/ipsetmanager.go

```diff
@@ -23,261 +25,346 @@ func (iMgr *IPSetManager) exists(name string) bool {
-		}
-
-		if list.Kind != ListSet {
-			return npmerrors.Errorf(npmerrors.AppendIPSet, false, fmt.Sprintf("ipset %s is not a list set", listName))
+			return npmerrors.Errorf(npmErrorString, false, fmt.Sprintf("ipset %s is not a hash set", setName))
 		}
+	}
+	return nil
+}
 
-		// check if Set is a member of List
-		listSet, exists := list.MemberIPSets[setName]
-		if exists {
-			if listSet == set {
-				// Set is already a member of List
-				return nil
-			}
-			// Update the ipset in list
-			list.MemberIPSets[setName] = set
-			return nil
-		}
+func (iMgr *IPSetManager) AddToList(listName string, setNames []string) error {
+	iMgr.Lock()
+	defer iMgr.Unlock()
 
-		// update the Ipset member list of list
-		list.AddMemberIPSet(set)
-		set.IncIpsetReferCount()
-		// Update metrics of the IpSet
-		metrics.AddEntryToIPSet(listName)
+	if err := iMgr.checkForAddToListErrors(listName, setNames); err != nil {
+		return err
 	}
 
-	iMgr.updateDirtyCache(listName)
-
+	for _, setName := range setNames {
+		iMgr.addMemberIPSet(listName, setName)
+	}
+	iMgr.modifyCacheForKernelMemberUpdate(listName, true)
+	metrics.AddEntryToIPSet(listName)
 	return nil
 }
 
 func (iMgr *IPSetManager) RemoveFromList(listName string, setNames []string) error {
 	iMgr.Lock()
 	defer iMgr.Unlock()
+
+	if err := iMgr.checkForRemoveFromListErrors(listName, setNames); err != nil {
+		return err
+	}
+
 	for _, setName := range setNames {
-		set, exists := iMgr.setMap[setName] // check if the Set exists
-		if !exists {
-			return npmerrors.Errorf(npmerrors.DeleteIPSet, false, fmt.Sprintf("ipset %s does not exist", setName))
-		}
+		iMgr.removeMemberIPSet(listName, setName)
+	}
+	iMgr.modifyCacheForKernelMemberUpdate(listName, false)
+	metrics.RemoveEntryFromIPSet(listName)
+	return nil
+}
 
-		if set.Kind != HashSet {
-			return npmerrors.Errorf(npmerrors.DeleteIPSet, false, fmt.Sprintf("ipset %s is not a hash set", setName))
-		}
+func (iMgr *IPSetManager) checkForAddToListErrors(listName string, setNames []string) error {
+	if err := iMgr.checkForListMemberUpdateErrors(listName, setNames, npmerrors.AppendIPSet); err != nil {
+		return err
+	}
 
-		// Nested IPSets are only supported for windows
-		// Check if we want to actually use that support
-		if set.Kind != HashSet {
-			return npmerrors.Errorf(npmerrors.DeleteIPSet, false, fmt.Sprintf("member ipset %s is not a Set type and nestetd ipsets are not supported", setName))
+	list := iMgr.setMap[listName]
+	for _, setName := range setNames {
+		if list.hasMember(setName) {
+			return npmerrors.Errorf(npmerrors.AppendIPSet, false, fmt.Sprintf("ipset %s is already a member of ipset %s", setName, listName))
 		}
+	}
+	return nil
+}
 
-		list, exists := iMgr.setMap[listName] // check if the Set exists
-		if !exists {
-			return npmerrors.Errorf(npmerrors.DeleteIPSet, false, fmt.Sprintf("ipset %s does not exist", listName))
+func (iMgr *IPSetManager) checkForRemoveFromListErrors(listName string, setNames []string) error {
```

Comment:
lets remove this function and also checkForAddToListErrors, then can be part of their respective callers